File: Algorithms/Medium/725.Split-Linked-List-In-Parts/SplitLinkedListInParts.go
package medium

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time complexity: O(n)
// Space complexity: O(1)
func splitListToParts(head *ListNode, k int) []*ListNode {
	parts := make([]*ListNode, 0, k)

	curr, length := head, 0
	for curr != nil {
		length++
		curr = curr.Next
	}

	subLength := length / k
	reminders := length % k

	var prev *ListNode
	curr = head
	for ; k > 0; k-- {
		parts = append(parts, curr)

		skip := subLength
		if reminders > 0 {
			skip++
			reminders--
		}

		for ; skip > 0; skip-- {
			prev = curr
			curr = curr.Next
		}

		if prev != nil {
			prev.Next = nil
		}
	}

	return parts
}

File: Algorithms/Medium/725.Split-Linked-List-In-Parts/SplitLinkedListInParts.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# Time complexity: O(k + n)
# Space complexity: O(1)
class Solution:
    def splitListToParts(
        self, head: Optional[ListNode], k: int
    ) -> list[Optional[ListNode]]:
        result = [None] * k

        length = self.listLen(head)
        one_more = length % k

        curr = head
        for i in range(k):
            skip = length // k - 1
            if one_more > 0:
                skip += 1
                one_more -= 1

            for _ in range(skip):
                curr = curr.next

            if head:
                result[i] = head
                head = curr.next
                curr.next = None
                curr = head

        return result

    def listLen(self, head: Optional[ListNode]) -> int:
        n, curr = 0, head
        while curr:
            n += 1
            curr = curr.next

        return n
